pkg/registry/registry.go

```diff
@@ -0,0 +1,239 @@
+/*
+Copyright (c) 2016-2017 Bitnami
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package registry
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+	"regexp"
+	"time"
+
+	"k8s.io/api/core/v1"
+)
+
+// Credentials represent the required credentials to authenticate against a Docker registry
+type Credentials struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+	Email    string `json:"email,omitifempty"`
+	Auth     string `json:"auth,omitifempty"`
+}
+
+// Registry struct represents a Docker Registry
+type Registry struct {
+	Endpoint string
+	Version  string
+	Creds    Credentials
+}
+
+type tagv1 struct {
+	Layer string `json:"layer"`
+	Name  string `json:"name"`
+}
+
+type tagListV2 struct {
+	Name string   `json:"name"`
+	Tags []string `json:"tags"`
+}
+
+type dockerCfg struct {
+	Auths map[string]Credentials `json:"auths"`
+}
+
+// New returns a Registry struct parsing its URL and storing the required credentials
+func New(config v1.Secret) (*Registry, error) {
+	// Parse secret
+	cfg := dockerCfg{}
+	err := json.Unmarshal(config.Data[".dockerconfigjson"], &cfg)
```

Comment:
Should that be `dockercfg`?

``` console
kubectl get secret kubeless-registry-credentials  --output=json
{
    "apiVersion": "v1",
    "data": {
        ".dockercfg": "eyJhdXRocyI6eyJodHRwczovL2luZGV4LmRvY2tlci5pby92MS8iOnsidXNlcm5hbWUiOiJtdXJhbGlyZWRkeSIsInBhc3N3b3JkIjoiTmV3bGlmZX4xIiwiZW1haWwiOiJtdXJhbGltbXJlZGR5QGdtYWlsLmNvbSIsImF1dGgiOiJiWFZ5WVd4cGNtVmtaSGs2VG1WM2JHbG1aWDR4In19fQ=="
    },
```

Reply:
That's the old output. That is the warning I put in the doc, it is required to use `kubectl` 1.9+ in order to generate the correct secret: https://github.com/kubernetes/kubernetes/pull/57463

Reply:
I see. Some how i am still getting old format.
```
± k version
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.0", GitCommit:"925c127ec6b946659ad0fd596fa959be43f0cc05", GitTreeState:"clean", BuildDate:"2017-12-15T21:07:38Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"", Minor:"", GitVersion:"v1.9.0", GitCommit:"925c127ec6b946659ad0fd596fa959be43f0cc05", GitTreeState:"clean", BuildDate:"2018-01-26T19:04:38Z", GoVersion:"go1.9.1", Compiler:"gc", Platform:"linux/amd64"}

± k get secret kubeless-registry-credentials -o yaml
apiVersion: v1
data:
  .dockercfg:
```

Reply:
mm, not sure about which specific version fixes it, the latest one has the fix:
```
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-07T12:22:21Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"darwin/amd64"}
```

Reply:
Used 1.9.3, its in the format as expected.